// Code generated by github.com/tizz98/results; DO NOT EDIT.
package results

import (
	"context"
	"fmt"
	"time"
)

type TimePtrResult struct {
	value **time.Time
	err   error
}

// SetNewTimePtrResult is a shortcut to creating a new TimePtrResult and then calling .Set(v, err) on it.
func SetNewTimePtrResult(v *time.Time, err error) (result TimePtrResult) {
	result.Set(v, err)
	return
}

// IsOk returns true when the result contains a non-nil result with no error
func (r TimePtrResult) IsOk() bool {
	return r.err == nil
}

// IsErr returns true when the result contains a non-nil error
func (r TimePtrResult) IsErr() bool {
	return r.err != nil
}

// Unwrap panics if the result contains an error, otherwise it returns the value
func (r TimePtrResult) Unwrap() *time.Time {
	if r.IsErr() {
		panic("cannot unwrap TimePtrResult, it is an error")
	}
	return *r.value
}

// Expect panics with the specified message if the result contains an error, otherwise it returns the value
func (r TimePtrResult) Expect(message string) *time.Time {
	if r.IsErr() {
		panic(fmt.Errorf("%s: %w", message, r.GetErr()))
	}
	return *r.value
}

// Expectf panics with the specified message if the result contains an error, otherwise it returns the value.
// This is different than Expect because if will automatically format the string with the given args.
func (r TimePtrResult) Expectf(format string, args ...interface{}) *time.Time {
	if r.IsErr() {
		panic(fmt.Errorf("%s: %w", fmt.Sprintf(format, args...), r.GetErr()))
	}
	return *r.value
}

// UnwrapOr returns the value if there is not an error, otherwise the specified value is returned
func (r TimePtrResult) UnwrapOr(v *time.Time) *time.Time {
	if r.IsOk() {
		return r.Unwrap()
	}
	return v
}

// UnwrapOrElse returns the value if there is not an error, otherwise the function is called and the result is returned
func (r TimePtrResult) UnwrapOrElse(fn func(err error) *time.Time) *time.Time {
	if r.IsOk() {
		return r.Unwrap()
	}
	return fn(r.err)
}

// Ok sets the result to a successful result with the provided value.
// This will panic if the result has already been set to successful or an error.
func (r *TimePtrResult) Ok(v *time.Time) {
	r.checkAbilityToSet()
	r.value = &v
}

// Err sets the result to an error result with the provided error.
// This will panic if the result has already been set to successful or an error.
func (r *TimePtrResult) Err(err error) {
	r.checkAbilityToSet()
	r.err = err
}

// GetError returns the error of the result. It may be nil, so check with TimePtrResult.IsErr() first.
func (r TimePtrResult) GetErr() error {
	return r.err
}

// Tup returns a tuple of (*time.Time, error) with nil being returned for *time.Time if there is an error
func (r TimePtrResult) Tup() (*time.Time, error) {
	return r.UnwrapOr(nil), r.err
}

// Set is a shortcut to checking the value of an error before setting the result.
// If there is an error, TimePtrResult.Err(err) will be called, otherwise TimePtrResult.Ok(v) will be called.
func (r *TimePtrResult) Set(v *time.Time, err error) {
	if err != nil {
		r.Err(err)
		return
	}

	r.Ok(v)
}

func (r TimePtrResult) checkAbilityToSet() {
	if r.isSet() {
		panic("TimePtrResult is already set, cannot set again")
	}
}

func (r TimePtrResult) isSet() bool {
	return r.value != nil || r.err != nil
}

// ContextWithTimePtr embeds the given value of *time.Time into the context for later retrieval with TimePtrFromContext
func ContextWithTimePtr(ctx context.Context, key interface{}, v *time.Time) context.Context {
	return context.WithValue(ctx, key, v)
}

// TimePtrFromContext attempts to retrieve a *time.Time value from the specified context. A TimePtrResult is returned
// which can be used to inspect the success or failure of retrieval.
func TimePtrFromContext(ctx context.Context, key interface{}) (result TimePtrResult) {
	if v, ok := ctx.Value(key).(*time.Time); !ok {
		result.Err(fmt.Errorf("%#v not found in context", key))
	} else {
		result.Ok(v)
	}
	return
}
